pandas/core/case_when.py

```diff
@@ -0,0 +1,151 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from pandas.core.dtypes.cast import (
+    construct_1d_arraylike_from_scalar,
+    find_common_type,
+    infer_dtype_from,
+)
+from pandas.core.dtypes.common import is_scalar
+from pandas.core.dtypes.generic import ABCSeries
+
+from pandas.core.construction import array as pd_array
+
+if TYPE_CHECKING:
+    from pandas._typing import (
+        ListLike,
+        Scalar,
+        Series,
+    )
+
+
+def case_when(
+    *args: ListLike | Scalar,
```

Comment:
I'm not a big fan of this signature. How about `conditions: ListLike | Scalar, replacements: ListLike | Scalar` and `conditions` and `replacements` will need to be validated to be both scalars or equal length

Reply:
Using *args allows the `condition, replacement` logic, instead of creating a list of conditions and a list of replacements. It looks cleaner and more understandable - each condition is immediately followed by its supposed replacement.

Reply:
I don't completely agree.

1. I would imagine once there are +5 arguments it's not immediately parse-able which arguments are conditions and which ones are replacements
2. Having separate conditions/replacements arguments avoids a potential footgun when `replacements` is/are also boolean values. In an `*arg` signature, a boolean replacement could be mistaken as a condition leading to silently unintended behavior

Reply:
@mroeschke how is that mistake possible? Is there an example you can share? In SQL this is how it is expressed, `when condition then replacement,...`. An example is likely to help me understand

Reply:
```python
df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]))
pd.case_when(df.a == 1, df.b != 4, df.a < 1, df.a == 3, df.b > 3, df.b < 6)
```

With the current signature, I think this is a valid use case and could be hard to discover if condtions vs replacements are intended

Reply:
@mroeschke thanks for the example provided. if it is rearranged, it becomes easier to grok: 
```py
pd.case_when(df.a == 1, df.b != 4,   # condition, replacement
             df.a < 1, df.a == 3, 
             df.b > 3, df.b < 6)
```

The condition is followed immediately by its replacement, instead of putting them in separate sequences. This way it is similar to SQL's case_when: 

```py
case when a = 1 then b != 4
     when a < 1 then b = 3
     when b > 3 then b < 6
```

same thing in R's data.table and python's datatable and dplyr (inspiration for this form is from these packages). The idea is that replacement should immediately follow condition, which i believe makes sense. a better option might be to just allow `mask/where` support multiple conditions and assignments, instead of creating a new function. hoping others would chime in as well

Reply:
@samukweku - those using auto formatters such as black will have to turn off and then on automatic formatting in order to achieve the format you show.

What if we asked for a list of tuples instead?

    pd.case_when((df.a == 1, df.b != 4), (df.a < 1, df.a == 3), (df.b > 3, df.b < 6))

This would also be okay with line-by-line formatting:

```
pd.case_when(
	(df.a == 1, df.b != 4), 
	(df.a < 1, df.a == 3), 
	(df.b > 3, df.b < 6),
)
```

Reply:
My bad, didn't think about linters. I am open to either option - separate arguments for conditions and replacements, or variable args of tuples.

Reply:
Requiring args to be tuples would be a good middle ground - can hopefully be better typed in the signature as well